### PR TITLE
G2P-29 - Enter Marsh mechanism probabilities

### PR DESCRIPTION
**Pull request by olaaustine**
Script to input Marsh mechanism probabilities

### CODE REVIEW

File: scripts/import/probabilities/marsh_probabilities.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+            list_lines[i] = line
+        
+    cursor.close()
+    database.close()
+
+
+    return list_lines
+
+
+def get_hgnc_id_from_g2p_db(list_lines, host, port, db, password, user):
```
Comment (dglemos) [inline scripts/import/probabilities/marsh_probabilities.py:93]: I don't think this subroutine is necessary.
The script already linked the score to the gene ID, this subroutine is only using that existing link to create a link to the HGNC.

File: scripts/import/probabilities/marsh_probabilities.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        print("The type applied is not permitted")
+        sys.exit()
+
+    if file:
+        print("Getting details from file")
+        file_lines = get_details_from_file(file)
+        print("Getting locus id from the G2P DB")
+        get_locus_id_from_g2p_db(file_lines, host, port, db, pwd, user)
+        print("Getting HGNC locus identifier from G2P DB")
+        get_hgnc_id_from_g2p_db(file_lines, host, port, db, pwd, user)
```
[the remark this answers is not in the document]
Comment (olaaustine) [reply scripts/import/probabilities/marsh_probabilities.py:313]: What function and what do you expect it to save?

File: scripts/import/probabilities/marsh_probabilities.py (hunk 1 above)
Comment (olaaustine) [reply scripts/import/probabilities/marsh_probabilities.py:95]: Yeah i agree, removed HGNC (locus_identifier) from the gene_stats table

File: scripts/import/probabilities/marsh_probabilities.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    get_locus_source = """ SELECT id from locus where name = %s
+"""
+
+    database = MySQLdb.connect(host=host,port=port,user=user,passwd=password,db=db)
+    cursor = database.cursor()
+
+    for i, line in enumerate(list_lines):
+        if isinstance(line, str):
+            line = line.split()
+            gene_symbol = line[0]
```
[the remark this answers is not in the document]
Comment (olaaustine) [reply scripts/import/probabilities/marsh_probabilities.py:75]: Added the check to the locus_attrib also

[the remark this answers is not in the document]
Comment (olaaustine) [reply scripts/import/probabilities/marsh_probabilities.py:76]: Added this check also

File: scripts/import/probabilities/marsh_probabilities.py [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+        - This script requires the `attrib_mapping` dictionary to be predefined, which maps attribute types to description IDs.
+    """
+
+    parser = argparse.ArgumentParser(description="This script is used to import the probabilities from a file and imports it to the gene_stats table in the G2P DB")
+    parser.add_argument("--host", required=True, help="Host of the database were the data is to be imported")
+    parser.add_argument("-p", "--port", required=True, help="Port information of the database to be imported")
+    parser.add_argument("-d", "--database", required=True, help="G2P Database to import the information into")
+    parser.add_argument("-pwd", "--password", required=True, help="Paaword for the G2P database information")
+    parser.add_argument("-u", "--user", required=True, help="User information for the G2P database")
+    parser.add_argument("-f", "--file", required=True, help="File containing the information for the score, can either be one file or files seperated by ,")
```
Comment (dglemos) [inline scripts/import/probabilities/marsh_probabilities.py:409]: > can either be one file or files seperated by ,

The script is only expecting one file.

File: scripts/import/probabilities/marsh_probabilities.py [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+        - Ensure that the database and tables are correctly configured before running the script.
+        - The file provided should follow the format expected by the script for proper processing.
+        - This script requires the `attrib_mapping` dictionary to be predefined, which maps attribute types to description IDs.
+    """
+
+    parser = argparse.ArgumentParser(description="This script is used to import the probabilities from a file and imports it to the gene_stats table in the G2P DB")
+    parser.add_argument("--host", required=True, help="Host of the database were the data is to be imported")
+    parser.add_argument("-p", "--port", required=True, help="Port information of the database to be imported")
+    parser.add_argument("-d", "--database", required=True, help="G2P Database to import the information into")
+    parser.add_argument("-pwd", "--password", required=True, help="Paaword for the G2P database information")
```
Comment (dglemos) [inline scripts/import/probabilities/marsh_probabilities.py:407]: Paaword -> Password

File: scripts/import/probabilities/marsh_probabilities.py [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+    """
+
+    get_source_query = """ SELECT id from source where name = 'Marsh Mechanism probabilities'"""
+
+    database = MySQLdb.connect(host=host,port=port,user=user,passwd=password,db=db)
+    cursor = database.cursor()
+
+    cursor.execute(get_source_query)
+    source_id = cursor.fetchone()
+    source_id = source_id[0]
```
Comment (dglemos) [inline scripts/import/probabilities/marsh_probabilities.py:264]: Before fetching the results, this method should fail if the source cannot be found.

File: scripts/import/probabilities/marsh_probabilities.py [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+    insert_into_gene_stats_query = """ INSERT into gene_stats (gene_symbol, gene_id, score, source_id, description_id) VALUES (%s, %s, %s, %s, %s)
+"""
+    source_id = get_source_details(host, port, db, password, user)
+
+    attrib_value = get_attrib_ids(host,port,db,password,user,attrib)
+
+    database = MySQLdb.connect(host=host,port=port,user=user,passwd=password,db=db)
+    cursor = database.cursor()
+
+
```
Comment (dglemos) [inline scripts/import/probabilities/marsh_probabilities.py:342]: Here you can print
`print("Genes skipped:")`

**Review by dglemos — COMMENTED** [review]
The script works, but in the future it could be improved to only iterate the file once.
The method `insert_into_gene_stats` could fetch the locus_id and then insert the score or skip the line, `get_locus_id_from_g2p_db` is not necessary.

**Review by dglemos — APPROVED** [review]

**Pull request merged**